src/main/java/ch/njol/skript/effects/EffSort.java

```diff
@@ -0,0 +1,167 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.effects;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Keywords;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.ExprInput;
+import ch.njol.skript.expressions.ExprSortedList;
+import ch.njol.skript.lang.Effect;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.InputSource;
+import ch.njol.skript.lang.ParseContext;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.Variable;
+import ch.njol.skript.lang.parser.ParserInstance;
+import ch.njol.util.Kleenean;
+import ch.njol.util.Pair;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+@Name("Sort")
+@Description({
+	"Sorts a list variable using either the natural ordering of the contents or the results of the given expression.",
+	"Be warned, this will overwrite the indices of the list variable."
+})
+@Examples({
+	"set {_words::*} to \"pineapple\", \"banana\", \"yoghurt\", and \"apple\"",
+	"sort {_words::*} # alphabetical sort",
+	"sort {_words::*} by length of input # shortest to longest",
+	"sort {_words::*} based on {tastiness::%input%} # sort based on custom value"
+})
+@Since("INSERT VERSION")
+@Keywords("input")
+public class EffSort extends Effect implements InputSource {
+
+	static {
+		Skript.registerEffect(EffSort.class, "sort %~objects% [(by|based on) <.+>]");
+		if (!ParserInstance.isRegistered(InputData.class))
+			ParserInstance.registerData(InputData.class, InputData::new);
+	}
+
+	@Nullable
+	private Expression<?> mappingExpr;
+	@Nullable
+	private String unparsedExpression;
+	private Variable<?> unsortedObjects;
+
+	private Set<ExprInput<?>> dependentInputs = new HashSet<>();
+
+	@Nullable
+	private Object currentValue;
+	@UnknownNullability
+	private String currentIndex;
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (expressions[0].isSingle() || !(expressions[0] instanceof Variable)) {
+			Skript.error("You can only sort list variables!");
+			return false;
+		}
+		unsortedObjects = (Variable<?>) expressions[0];
+
+		if (parseResult.regexes.size() > 0) {
```

Comment:
```suggestion
		if (!parseResult.regexes.isEmpty()) {
```
this might be a little clearer